### PR TITLE
Allow for more CORS configuration

### PR DESCRIPTION
**Pull request by glbrntt**
Motivation:

We added some level of CORS configuration support in #1583. This change adds further flexibility.

Modifications:

- Add an 'originBased' mode where the value of the origin header is returned in the response head.
- Add a custom fallback where the user can specify a callback which is passed the value of the origin header and returns the value to return in the 'access-control-allow-origin' response header (or nil, if the origin is not allowed).

Result:

More flexibility for CORS.

### CODE REVIEW

File: Sources/GRPC/Server.swift [hunk 1]
```diff
@@ -490,10 +527,25 @@ extension Server.Configuration.CORS {
...
     }
+
+    /// Provide a custom CORS origin check.
+    ///
+    /// - Parameter checkOrigin: A closure which is called with the value of the 'origin' header
+    ///     and returns the value to use in the 'access-control-allow-origin' response header,
+    ///     or `nil` if the origin is not allowed.
+    public static func custom(
+      _ checkOrigin: @escaping @Sendable (_ origin: String) -> String?
+    ) -> Self {
```
Comment (Lukasa) [inline Sources/GRPC/Server.swift:546]: Given the requirement for sendability and hashability, should this take a generic type or something equivalent to enable you to maintain the invariant?

Comment (glbrntt) [reply Sources/GRPC/Server.swift:546]: Hmm, we'd need to store it as an existential which means we'd get the automatic `Hashable` conformance but not the `Equatable` conformance (the stored `any P`s could be of different types). It's also mildly inconvenient for callers who are implementing that type. My lean is towards not making this generic.

Comment (Lukasa) [reply Sources/GRPC/Server.swift:546]: We can get the equatable and hashable conformances using `AnyHashable` and more generally can achieve all of this using closures. The question is whether it's a good idea to keep types that don't really behave the way we want them to, I think.

File: Sources/GRPC/Server.swift [hunk 2]
```diff
@@ -520,3 +535,46 @@ extension Comparable {
...
+  ///     CORS related headers should be returned.
+  func check(origin: String) -> String?
+}
+
+extension Server.Configuration.CORS.AllowedOrigins {
+  struct AnyCustomCORSAllowedOrigin: GRPCCustomCORSAllowedOrigin {
+    private var checkOrigin: @Sendable (String) -> String?
+    private let hashInto: @Sendable (inout Hasher) -> Void
+    #if swift(>=5.7)
+    private let isEqualTo: @Sendable (any GRPCCustomCORSAllowedOrigin) -> Bool
```
Comment (dnadoba) [inline Sources/GRPC/Server.swift:554]: Do we get warnings if we don't use `any` with new swift compilers?

**Review by Lukasa — APPROVED** [review]

**Pull request merged**